### PR TITLE
Add vn search

### PR DESCRIPTION
**Pull request by Moltenship**

### CODE REVIEW

**Review by moetazaneta — APPROVED** [review]

File: src/api/mappers/paginationMapper.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import type { z } from 'zod';
+import type { PaginationDto } from '../dtos/paginationDto';
+import type { Pagination } from '../models/pagination';
+
+type MapperFn<T extends z.ZodTypeAny, R> = (data: z.infer<T>) => R;
+
+export namespace PaginationMapper {
+
+  /**
+   *
```
Comment (moetazaneta) [inline src/api/mappers/paginationMapper.ts]: Redundant line

File: src/api/models/platform.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+   */
+  export function toReadable(platform: Platform): string {
+    return PLATFORM_READABLE_MAP[platform] ?? 'Other platform';
+  }
+
+  /**
+   * Gets flag icon name.
+   * @param platform Language code (ISO-639).
+   */
+  export function getPlatformIconName(platform: Platform): PlatformIconName {
```
Comment (moetazaneta) [inline src/api/models/platform.ts]: Do we really need to repeat `Platform` in method name? I suggest `getIconName` for all such models

File: src/api/mappers/tagMapper.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  /**
+   * Maps dto to model.
+   * @param dto Tag dto.
+   */
+  export function fromDto(dto: TagDto): Tag {
+    return {
+      id: dto.id,
+      aliases: dto.aliases,
+      category: dto.category,
```
Comment (moetazaneta) [inline src/api/mappers/tagMapper.ts]: Maybe we should add mapper for category, what do you think?
We have mappers for `VnDevelopmentStatus` and `VnLength`. The only difference I see is that `VnLength` is union of numbers and `TagCategory` is union of strings.

File: src/api/models/vn/length.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+export namespace VnLength {
+
+  /**
+   * Converts length to a readable equivalent.
+   * @param value Vn length.
+   */
+  export function toReadable(value: VnLength | null): string {
+    const UNKNOWN_LENGTH = 'Unknown';
+    return value !== null ? LENGTH_READABLE_MAP[value] ?? UNKNOWN_LENGTH : UNKNOWN_LENGTH;
```
Comment (moetazaneta) [inline src/api/models/vn/length.ts:21]: We could extract `"Unknown"` into global constant.

File: src/api/models/vn/vn.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  readonly aliases: readonly string[];
+
+  /** Original language for vn. */
+  readonly originalLanguage: LanguageCode;
+
+  /** Current dev status of vn. */
+  readonly developmentStatus: VnDevelopmentStatus;
+
+  /** Release date. */
+  readonly released: string | null;
```
Comment (moetazaneta) [inline src/api/models/vn/vn.ts:35]: We have difference logic for `released` field. In search query `released` means range of numbers, here it's string.
First of all, I suggest using `Date`, and second, differenciate fields, to better convey  the meaning.

File: src/api/models/vn/vn.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  readonly rating: number | null;
+
+  /** Popularity(0-100). */
+  readonly popularity: number;
+
+  /** Votes. */
+  readonly voteCount: number;
+
+  /** Screenshots. */
+  readonly screenshots: VnScreenshot[];
```
Comment (moetazaneta) [inline src/api/models/vn/vn.ts]:
```suggestion
  readonly screenshots: readonly VnScreenshot[];
```

File: src/api/models/queryBody.ts [hunk 7]
```diff
@@ -10,7 +10,7 @@ export interface QueryBody<
   readonly count: boolean;
```
Comment (moetazaneta) [inline src/api/models/queryBody.ts:10]: Typo in doc `/** Whether to count **or** items that satisfy given query or not. */`.
Also property name is unclear, I suggest naming it 'isCountItems' or maybe `doCountItems`.

File: src/api/utils/isEmpty.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Check if values is not empty.
+ * @param value Value.
+ */
+export function isNotEmpty<T>(value: T): value is NonNullable<T> {
```
Comment (moetazaneta) [inline src/api/utils/isEmpty.ts]: Rename file

File: src/components/Field/Field.tsx [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import type { ComponentProps, ElementType } from 'react';
+import { useController } from 'react-hook-form';
+import type { FieldValues } from 'react-hook-form';
+
+import type { FormControlProps } from 'src/utils/FormControlProps';
+
+type Props<C extends ElementType, T extends FieldValues> = {
+  readonly Component: C;
+
+} & ComponentProps<C> & Omit<ComponentProps<C>, 'name'> & FormControlProps<T>;
```
Comment (moetazaneta) [inline src/components/Field/Field.tsx]:
```suggestion
type Props<C extends ElementType, T extends FieldValues> = 
& {
  readonly Component: C;
} 
& ComponentProps<C> 
& Omit<ComponentProps<C>, 'name'> 
& FormControlProps<T>;
```

File: src/components/Input/Input.tsx [hunk 10]
```diff
@@ -14,18 +14,21 @@ export interface InputProps {
...
-  readonly name: string;
+  readonly name?: string;
 
   /** Input value. */
-  readonly value: string;
+  readonly value?: string;
 
   /** Input change callback. */
-  readonly onChange: ChangeEventHandler<HTMLInputElement>;
+  readonly onChange?: ChangeEventHandler<HTMLInputElement>;
```
Comment (moetazaneta) [inline src/components/Input/Input.tsx]: Doesn't `InputHTMLAttributes<HTMLInputElement>` have `onChange` prop? I think we can omit it, also probably `value` and `name` too.

File: src/components/Input/Input.tsx [hunk 11]
```diff
@@ -63,16 +76,22 @@ const InputComponent = forwardRef<HTMLInputElement, Props>(({
...
-        className="grow rounded-md border-none bg-inherit py-3  pl-4 pr-2 text-sm leading-6 focus:outline-none"
+        onFocus={e => {
+          setIsInputGroupFocused(true);
+          onFocus?.(e);
+        }}
+        onBlur={e => {
+          setIsInputGroupFocused(false);
+          onBlur?.(e);
+        }}
+        className={clsx(`grow rounded-md border-none bg-inherit py-3 ${leftElement ? 'pl-12' : 'pl-4'} ${leftElement ? 'pr-12' : 'pr-4'} text-sm leading-6 focus:outline-none`, className)}
```
Comment (moetazaneta) [inline src/components/Input/Input.tsx]:
```suggestion
        className={clsx(
                'grow rounded-md border-none bg-inherit py-3 text-sm leading-6 focus:outline-none', className, {
                        'px-12': leftElement,
                        'px-4': !leftElement,
                }
        )}
```

File: src/components/Popover/Popover.tsx [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import * as RadixPopover from '@radix-ui/react-popover';
+import clsx from 'clsx';
+import type { FC, ForwardedRef } from 'react';
+import { forwardRef, memo } from 'react';
+import { inter } from 'src/pages/_app';
```
Comment (moetazaneta) [inline src/components/Popover/Popover.tsx:5]: Lets rename it to `font` or something else abstract.

Also did you try using `inherit`?

File: src/components/Select/Select.tsx [hunk 13]
```diff
@@ -152,4 +160,11 @@ const SelectComponent = <
   );
 };
 
-export const Select = typedMemo(SelectComponent);
+declare module 'react' {
```
Comment (moetazaneta) [inline src/components/Select/Select.tsx:164]: Please leave explaining comment

File: src/components/Skeleton/Skeleton.tsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import clsx from 'clsx';
+import type { FC } from 'react';
+import { memo } from 'react';
+import type { PropsWithChildrenAndClass } from 'src/utils/PropsWithClass';
+
+/** Skeleton component. */
```
Comment (moetazaneta) [inline src/components/Skeleton/Skeleton.tsx:7]: Please add todo to "decompose this component into text and rectangle components after designs are ready".

File: src/features/search/components/SearchHeader/SearchHeader.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import { useRouter } from 'next/router';
+import type { FC } from 'react';
+import { useCallback } from 'react';
+import type { TabValue } from '../SearchHeaderTabs/SearchHeaderTabs';
+import { SearchHeaderTabs } from '../SearchHeaderTabs/SearchHeaderTabs';
+
+/** Header for search pages. */
+export const SearchHeader: FC = () => {
+  const router = useRouter();
+  const activeTabValue = router.route.split('/').at(-1) as TabValue;
```
Comment (moetazaneta) [inline src/features/search/components/SearchHeader/SearchHeader.tsx:10]: I think we should come up with solution for this (hook maybe?)

File: src/features/search/components/SearchHeader/SearchHeader.tsx [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+  const handleTabChange = useCallback((tabName: TabValue) => {
+    router.push({
+      pathname: `./${tabName}`,
+    });
+  }, []);
+
+  return (
+    <header className="mb-4">
```
Comment (moetazaneta) [inline src/features/search/components/SearchHeader/SearchHeader.tsx]: Please don't encapsulate margin in component, do it in parent component

File: src/features/search/components/SearchHeaderTabs/SearchHeaderTabs.tsx [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+const SearchHeaderTabsComponent: FC<Props> = ({
+  value,
+  onChange,
+}) => (
+  <Tabs.Root
+    asChild
+    activationMode="manual"
+    value={value}
+
+    // Making tabs think that it works with `strings`, but it actually works with `TabValue` union.
```
Comment (moetazaneta) [inline src/features/search/components/SearchHeaderTabs/SearchHeaderTabs.tsx:37]: :)

Comment (moetazaneta) [reply src/features/search/components/SearchHeaderTabs/SearchHeaderTabs.tsx:37]: I think we should create wrapper for tabs, that encapsulate such logic. It probably should take readable map object as input.
Please add todo for this.

File: src/features/search/components/VnSearch/VnSearch.tsx [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+import { VnSearchForm } from '../VnSearchForm/VnSearchForm';
+
+/** Visual novel overview tab. */
+const VnSearchComponent: FC = () => {
+  const [currentPage, setCurrentPage] = useState(1);
+
+  const methods = useForm({ defaultValues: VN_SEARCH_INITIAL_VALUES });
+  const { control } = methods;
+
+  const formData = useWatch({ control }) as VnSearchFormValues;
```
Comment (moetazaneta) [inline src/features/search/components/VnSearch/VnSearch.tsx:22]: I believe you can use `watch()` from `methods` instead of `useWatch()`. This way it'll have proper typings.

File: src/features/search/components/VnSearch/VnSearch.tsx [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+          <List size={DEFAULT_PAGE_SIZE}>
+            <CardSkeleton />
+          </List>
+        )}
+      </div>
+
+      {/* TODO: Add placeholder for empty response. */}
+      {vns?.pages.at(-1)?.count === 0 && (
+        <div className="flex flex-col items-center">
+          <h2>УВЫ</h2>
```
Comment (moetazaneta) [inline src/features/search/components/VnSearch/VnSearch.tsx:69]: image

File: src/features/search/components/VnSearchPopover/VnSearchPopover.tsx [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+import { IconButton } from 'src/components/IconButton/IconButton';
+import { Popover, PopoverContent, PopoverTrigger } from 'src/components/Popover/Popover';
+
+/** Popover for vn search. */
+const VnSearchPopoverComponent: FC<PropsWithChildren> = ({ children }) => (
+  <Popover>
+    <PopoverTrigger asChild>
+      <IconButton intent="tertiary" name="options" className="shrink-0" />
+    </PopoverTrigger>
+    <PopoverContent align="end" className="w-[560px] px-8 py-7">
```
Comment (moetazaneta) [inline src/features/search/components/VnSearchPopover/VnSearchPopover.tsx]:
```suggestion
    <PopoverContent align="end" className="w-max-[560px] w-full px-8 py-7">
```
Not sure if it will work, but it should

File: src/hooks/useDebounce.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
+import { useEffect, useState } from 'react';
+
+/**
+ * Debounces given value.
+ * @param value Value.
+ * @param delay Delay.
+ */
+export function useDebounce<T>(value: T, delay = 500): T {
```
Comment (moetazaneta) [inline src/hooks/useDebounce.ts]: Please lower it to 100 or at least 200

File: src/pages/search/vn.tsx [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+export const getServerSideProps: GetServerSideProps = async() => {
+  await queryClient.prefetchInfiniteQuery(getBaseTagsQueryOptions({}));
+  await queryClient.prefetchInfiniteQuery(
+    getBaseVnsQueryOptions(VnSearchFormValues.toQueryOptions()),
+  );
+
+  return {
+    props: {
+      // Workaround for https://github.com/TanStack/query/issues/1458
+      dehydratedState: JSON.parse(JSON.stringify(dehydrate(queryClient))),
```
Comment (moetazaneta) [inline src/pages/search/vn.tsx]: I think it's one more reason to make a helper function to compose return data.

**Review by moetazaneta — CHANGES_REQUESTED** [review]

File: src/api/models/vn/vn.ts (hunk 5 above)
Comment (Moltenship) [reply src/api/models/vn/vn.ts:35]: I want to avoid using dates in model due to `getServerSideProps` serialization.

File: src/components/Popover/Popover.tsx (hunk 12 above)
Comment (Moltenship) [reply src/components/Popover/Popover.tsx:5]: `Popover` is mounted into `body` via `Portal`, so `inherit` is not an option here

File: src/components/Skeleton/Skeleton.tsx (hunk 14 above)
Comment (Moltenship) [reply src/components/Skeleton/Skeleton.tsx:7]: I don't think we should decompose this component. It's better to add props for `text`/`rectangle` as other ui-kits do

File: src/features/search/components/VnSearch/VnSearch.tsx (hunk 18 above)
Comment (Moltenship) [reply src/features/search/components/VnSearch/VnSearch.tsx:22]: `useWatch` has less rerenders

File: src/api/mappers/paginationMapper.ts (hunk 1 above)
Comment (Moltenship) [reply src/api/mappers/paginationMapper.ts]: This is eslint setting

File: src/api/models/vn/vn.ts (hunk 5 above)
Comment (moetazaneta) [reply src/api/models/vn/vn.ts:35]: Okay. Please add note about date format then.

File: src/components/Skeleton/Skeleton.tsx (hunk 14 above)
Comment (moetazaneta) [reply src/components/Skeleton/Skeleton.tsx:7]: Good, then leave a todo comment about adding props after designs are ready

File: src/features/search/components/SearchHeaderTabs/SearchHeaderTabs.tsx (hunk 17 above)
Comment (Moltenship) [reply src/features/search/components/SearchHeaderTabs/SearchHeaderTabs.tsx:37]: Added todo

File: src/api/models/platform.ts [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+  xbo: 'xbox',
+  xxs: 'xbox',
+  mob: 'smartphone',
+  web: 'chrome',
+  vnd: 'question-mark',
+  oth: 'question-mark',
+};
+
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+export namespace Platform {
```
Comment (NightOfNights) [inline src/api/models/platform.ts]: I suppose we can turn off this rule for models

File: src/api/models/queryOptions/vn/vnSortField.ts [hunk 24]
```diff
@@ -0,0 +1,27 @@
+/** List of available sort fields for vn. */
+export const VN_SORT_FIELDS = ['id', 'title', 'released', 'popularity', 'rating', 'votecount', 'length'] as const;
```
Comment (NightOfNights) [inline src/api/models/queryOptions/vn/vnSortField.ts]: I don't think it would be useful to sort vns by id

File: src/api/services/paginationService.ts [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+   * Gets page data from infinite query results.
+   * @param data Infinite query results data.
+   * @param page Page.
+   */
+  export function getPageData<T>(data: InfiniteData<Pagination<T>>, page: number): T[] {
+    const pageParam = data.pageParams.indexOf(page);
+    const index = pageParam === -1 ? 0 : pageParam;
+    return data.pages[index]?.results as T[] ?? [];
+  }
+}
```
Comment (NightOfNights) [inline src/api/services/paginationService.ts]: I don't think we need such service, you can add this method to Pagination model

File: src/api/utils/isNotEmpty.ts [hunk 26]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Check if values is not empty.
+ * @param value Value.
+ */
+export function isNotEmpty<T>(value: T): value is NonNullable<T> {
+  if (value === undefined || value === null) {
+    return false;
+  }
```
Comment (NightOfNights) [inline src/api/utils/isNotEmpty.ts]:
```suggestion
  if (value == null) {
    return false;
  }
```

File: src/api/utils/isNotEmpty.ts [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+
+  if (typeof value === 'string' && value === '') {
+    return false;
+  }
```
Comment (NightOfNights) [inline src/api/utils/isNotEmpty.ts]:
```suggestion
  if (value === '') {
    return false;
  }
```

File: src/components/Input/Input.tsx [hunk 28]
```diff
@@ -14,37 +14,35 @@ export interface InputProps {
...
   placeholder,
   isDisabled,
   type,
+  leftElement,
+  onBlur,
   rightElement,
-}, ref) => {
+  onFocus,
+  className,
+}, ref: ForwardedRef<HTMLInputElement>) => {
```
Comment (NightOfNights) [inline src/components/Input/Input.tsx]: Since you extended props type, make sure that other attributes are also applied to the input. You can do it by using the spread notation `...props`

File: src/components/List/List.tsx [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+
+interface Props {
+
+  /** List size. */
+  readonly size: number;
+}
+
+/** Component renders children * size times. */
+const ListComponent: FC<PropsWithChildren<Props>> = ({ size, children }) => {
+  const items = [...Array(size).keys()].map(i => <Fragment key={i}>{children}</Fragment>);
```
Comment (NightOfNights) [inline src/components/List/List.tsx]: Should it be named `List` though? It seems it just duplicate its chidlren n times

File: src/components/controls/PasswordInput.tsx [hunk 30]
```diff
@@ -59,5 +59,3 @@ const PasswordInputComponent: FC<Props> = ({
     />
   );
 };
-
-export const PasswordInput = memo(PasswordInputComponent);
```
Comment (NightOfNights) [inline src/components/controls/PasswordInput.tsx:63]: Why did you remove `memo`?

File: src/components/controls/TextInput.tsx [hunk 31]
```diff
@@ -45,5 +48,3 @@ const TextInputComponent: FC<Props> = ({
     />
   );
 };
-
-export const TextInput = memo(TextInputComponent);
```
Comment (NightOfNights) [inline src/components/controls/TextInput.tsx:49]: Same

File: src/features/search/components/VnSearch/VnSearch.tsx [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+        </div>
+      )}
+
+      {vns?.pages[0]?.count != null && !isLoading && (
+        <div className="mt-auto flex w-full justify-center">
+          <Paginator
```
Comment (NightOfNights) [inline src/features/search/components/VnSearch/VnSearch.tsx]: Can we provide more convenient interface to work with? Instead of stuff like `vns?.pages[0]?.count != null` or `vns?.pages.at(-1)?.count === 0`

File: src/features/search/components/VnSearchForm/VnSearchForm.tsx [hunk 33]
```diff
@@ -0,0 +1,216 @@
...
+    .map(tag => ({ label: tag.name, value: tag.id })) ?? [];
+
+  const sortFieldOptions = VN_SORT_FIELDS
+    .map(field => ({ value: field, label: VnSortField.toReadable(field) }));
+
+  const devStatusOptions = VN_DEV_STATUSES
+    .map(status => ({ value: status, label: VnDevelopmentStatus.toReadable(status) }));
+
+  const lengthOptions = VN_LENGTHS
+    .map(length => ({ value: length, label: VnLength.toReadable(length) }));
```
Comment (NightOfNights) [inline src/features/search/components/VnSearchForm/VnSearchForm.tsx]: You can move this stuff out of the component's scope, since it does not rely on props/state

File: src/features/search/components/VnSearchForm/VnSearchForm.tsx [hunk 34]
```diff
@@ -0,0 +1,216 @@
...
+          />
+          <Field
+            Component={Select}
+            control={control}
+            options={sortFieldOptions}
+            closeMenuOnSelect
+            name="sortField"
+            className="rounded-r bg-gray-100"
+            disableSearch
+          />
```
Comment (NightOfNights) [inline src/features/search/components/VnSearchForm/VnSearchForm.tsx:204]: It seems that width of this select is based on length of the current label. Maybe we can improve it by making its width consistent regardless of label

File: src/features/search/components/VnSearchForm/VnSearchForm.tsx [hunk 35]
```diff
@@ -0,0 +1,216 @@
...
+            </div>
+            <div className="grid grid-cols-2 gap-4">
+              <Field
+                Component={Slider}
+                control={control}
+                name="released"
+                label="Release Date"
+                showValues
+                min={1980}
+                max={2023}
```
Comment (NightOfNights) [inline src/features/search/components/VnSearchForm/VnSearchForm.tsx]: Let's not hardcode the current year 🙃

File: src/features/search/queries/tag.ts [hunk 36]
```diff
@@ -0,0 +1,29 @@
+import type { UseInfiniteQueryOptions, UseInfiniteQueryResult } from '@tanstack/react-query';
+import { useInfiniteQuery } from '@tanstack/react-query';
+import type { Pagination } from 'src/api/models/pagination';
+import type { TagQueryOptions } from 'src/api/models/queryOptions/tag/tagQueryOptions';
+import type { Tag } from 'src/api/models/tag/tag';
+import { TagService } from 'src/api/services/tagService';
+
+/** Gets base options. */
+export const getBaseTagsQueryOptions = ({ search = '' }: TagQueryOptions): UseInfiniteQueryOptions<Pagination<Tag>, Error> => ({
```
Comment (NightOfNights) [inline src/features/search/queries/tag.ts:9]: Do we need to export it? Is it possible could be used somewhere else?

Comment (Moltenship) [reply src/features/search/queries/tag.ts:9]: It's used for prefetching tags in `getServerSideProps`

File: src/features/search/components/VnSearchForm/VnSearchForm.tsx (hunk 34 above)
Comment (Moltenship) [reply src/features/search/components/VnSearchForm/VnSearchForm.tsx:204]: This will be fixed by @moetazaneta with updated design

File: src/api/models/vn/length.ts (hunk 4 above)
Comment (Moltenship) [reply src/api/models/vn/length.ts:21]: There is no other `Unknow` for now, each model has it's own "default"

**Review by moetazaneta — APPROVED** [review]

File: src/api/models/vn/length.ts (hunk 4 above)
Comment (moetazaneta) [reply src/api/models/vn/length.ts:21]: @Moltenship It's true, yet all of them use "Unknown", we could use it as default unknown value for all models.

Comment (NightOfNights) [reply src/api/models/vn/length.ts:21]: I don't think we need to work with `Unknown`, better to throw an error right away

**Review by NightOfNights — APPROVED** [review]
